src/data/matrix/kronecker_product_bialgebra.lean

```diff
@@ -0,0 +1,198 @@
+
+lemma kronecker_map_smul_right [has_scalar R β] [has_scalar R γ] (f : α → β → γ)
+  (hf : ∀ (r : R) a b, f a (r • b) = r • f a b) (r : R)
+  (A : matrix l m α) (B : matrix n p β) :
+  kronecker_map f A (r • B) = r • kronecker_map f A B :=
+ext $ λ i j, hf _ _ _
+
+/-- When `f` is bilinear then `matrix.kronecker_map f` is also bilinear. -/
+@[simps]
+def kronecker_map_linear [comm_semiring R]
+  [add_comm_monoid α] [add_comm_monoid β] [add_comm_monoid γ]
+  [module R α] [module R β] [module R γ]
+  (f : α →ₗ[R] β →ₗ[R] γ) :
+  matrix l m α →ₗ[R] matrix n p β →ₗ[R] matrix (l × n) (m × p) γ :=
+linear_map.mk₂ R
+  (kronecker_map (λ r s, f r s))
+  (kronecker_map_add_left _ $ f.map_add₂)
+  (kronecker_map_smul_left _ $ f.map_smul₂)
+  (kronecker_map_add_right _ $ λ a, (f a).map_add)
+  (kronecker_map_smul_right _ $ λ r a, (f a).map_smul r)
+
+end
+
+
+variables {α R S β : Type*} [comm_semiring α] [comm_semiring R] [comm_semiring S] [comm_semiring β]
+variables [algebra α R] [algebra α S] [algebra α β] [algebra R β] [algebra S β]
+variables {l m n p l' m' n' p' : Type*}
+variables [fintype l] [fintype m] [fintype n] [fintype p]
+variables [fintype l'] [fintype m'] [fintype n'] [fintype p']
+
+variables (α β)
+
+-- TODO: move this
+def algebra.biprod [is_scalar_tower α R β] [is_scalar_tower α S β] :
+  R →ₗ[α] S →ₗ[α] β :=
+((algebra.lmul α β).to_linear_map.compl₂ ((algebra.linear_map S β).restrict_scalars α)).comp
+  ((algebra.linear_map R β).restrict_scalars α)
+
+@[simp]
+lemma algebra.biprod_apply [is_scalar_tower α R β] [is_scalar_tower α S β] (r : R) (s : S) :
+  algebra.biprod α β r s = algebra_map R β r * algebra_map S β s := rfl
+
+variables (R S β)
+
+@[simps]
+def kronecker_biprod [is_scalar_tower α R β] [is_scalar_tower α S β] :
+  (matrix l m R) →ₗ[α] (matrix n p S) →ₗ[α] matrix (l × n) (m × p) β :=
+kronecker_map_linear $ algebra.biprod α β
+
+lemma kronecker_biprod_reindex_left (eₗ : l ≃ l') (eₘ : m ≃ m') (A : matrix l m R)
+  (B : matrix n p S) [is_scalar_tower α R β] [is_scalar_tower α S β] :
+  kronecker_biprod α R S β (reindex_linear_equiv α R eₗ eₘ A) B =
+  reindex_linear_equiv α _ (eₗ.prod_congr (equiv.refl _)) (eₘ.prod_congr (equiv.refl _))
+  (kronecker_biprod α R S β A B) := by { ext ⟨i, i'⟩ ⟨j, j'⟩, refl }
+
+lemma kronecker_biprod_reindex_right (eₙ : n ≃ n') (eₚ : p ≃ p') (A : matrix l m R)
+  (B : matrix n p S) [is_scalar_tower α R β] [is_scalar_tower α S β] :
+  kronecker_biprod α R S β A (reindex_linear_equiv α S eₙ eₚ B) =
+  reindex_linear_equiv α _ ((equiv.refl _).prod_congr eₙ) ((equiv.refl _).prod_congr eₚ)
+  (kronecker_biprod α R S β A B) := by { ext ⟨i, i'⟩ ⟨j, j'⟩, refl }
+
+lemma kronecker_biprod_one_one [decidable_eq m] [decidable_eq n] [is_scalar_tower α R β]
+  [is_scalar_tower α S β] : kronecker_biprod α R S β (1 : matrix m m R)
+    (1 : matrix n n S) = (1 : matrix (m × n) (m × n) β) :=
+    by { ext ⟨i, i'⟩ ⟨j, j'⟩, simp [kronecker_biprod, one_apply, algebra_map_eq_smul_one, ite_smul,
+      ite_and] }
```

Comment:
This proof generalizes to:
```suggestion
lemma kronecker_biprod_one_one [decidable_eq m] [decidable_eq n] [is_scalar_tower α R β]
  [is_scalar_tower α S β] : kronecker_biprod α R S β (1 : matrix m m R)
    (1 : matrix n n S) = (1 : matrix (m × n) (m × n) β) :=
begin
  refine kronecker_map_one_one _ _ _ _,
  all_goals {simp}
end
```

after adding:
```lean
lemma kronecker_map_diagonal_diagonal [has_zero α] [has_zero β] [has_zero γ]
  [decidable_eq m] [decidable_eq n]
  (f : α → β → γ) (hf₁ : ∀ b, f 0 b = 0) (hf₂ : ∀ a, f a 0 = 0) (a : m → α) (b : n → β):
  kronecker_map f (diagonal a) (diagonal b) = diagonal (λ mn, f (a mn.1) (b mn.2)) :=
begin
  ext ⟨i₁, i₂⟩ ⟨j₁, j₂⟩,
  simp [diagonal, apply_ite f, ite_and, ite_apply, apply_ite (f (a i₁)), hf₁, hf₂],
end

lemma kronecker_map_one_one [has_zero α] [has_zero β] [has_zero γ]
  [has_one α] [has_one β] [has_one γ]
  [decidable_eq m] [decidable_eq n]
  (f : α → β → γ) (hf₁ : ∀ b, f 0 b = 0) (hf₂ : ∀ a, f a 0 = 0) (hf₃ : f 1 1 = 1) :
  kronecker_map f (1 : matrix m m α) (1 : matrix n n β) = 1 :=
(kronecker_map_diagonal_diagonal _ hf₁ hf₂ _ _).trans $ by simp only [hf₃, diagonal_one]
```

Reply:
Nice, applied!